subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderDependenciesTest.groovy

```diff
@@ -0,0 +1,188 @@
+import org.gradle.api.plugins.JavaPlugin
+import org.gradle.internal.service.DefaultServiceRegistry
+import org.gradle.jvm.tasks.Jar
+import org.gradle.plugins.ide.eclipse.EclipsePlugin
+import org.gradle.plugins.ide.eclipse.model.EclipseClasspath
+import org.gradle.plugins.ide.eclipse.model.ProjectDependency
+import org.gradle.plugins.ide.internal.configurer.EclipseModelAwareUniqueProjectNameProvider
+import org.gradle.plugins.ide.internal.tooling.EclipseModelBuilder
+import org.gradle.plugins.ide.internal.tooling.GradleProjectBuilder
+import org.gradle.test.fixtures.AbstractProjectBuilderSpec
+import org.gradle.test.fixtures.file.CleanupTestDirectory
+import org.gradle.testfixtures.ProjectBuilder
+import org.gradle.tooling.model.eclipse.EclipseRuntime
+import org.gradle.tooling.model.eclipse.EclipseWorkspaceProject
+import org.gradle.util.TestUtil
+import org.gradle.util.UsesNativeServices
+
+@UsesNativeServices
+@CleanupTestDirectory
+class EclipseModelBuilderDependenciesTest extends AbstractProjectBuilderSpec {
+    Project child1
+    Project child2
+    Project child3
+
+    def setup() {
+        project = TestUtil.builder(temporaryFolder.testDirectory).withName("project").build()
+        child1 = ProjectBuilder.builder().withName("child1").withParent(project).build()
+        child2 = ProjectBuilder.builder().withName("child2").withParent(project).build()
+        child3 = ProjectBuilder.builder().withName("child3").withParent(project).build()
+
+        def libsDir = new File(project.projectDir, "libs")
+        libsDir.mkdirs()
+        new File(libsDir, "test-1.0.jar").createNewFile()
+        [project, child1, child2, child3].each { it.pluginManager.apply(EclipsePlugin) }
+        [child1, child2, child3].each {
+            it.plugins.apply(JavaPlugin)
+            it.repositories.flatDir {
+                dirs libsDir
+            }
+        }
+
+        child1.configurations.create("testArtifacts")
+        def task = child1.tasks.create("testJar", Jar) {
+            classifier = "tests"
+            from(project.sourceSets.test.output)
+        }
+        child1.dependencies {
+            implementation "fakegroup:test:1.0"
+        }
+        child1.artifacts.add("testArtifacts", task)
+        child2.dependencies {
+            implementation child2.dependencies.project(path: ":child1")
+            testImplementation child2.dependencies.project(path: ":child1", configuration: "testArtifacts")
+        }
+        child3.eclipse {
+            (classpath as EclipseClasspath).file.whenMerged {
+                def customDependency = new ProjectDependency("/child1")
+                customDependency.buildDependencies(":child1:jar")
+                customDependency.publication = fileReference("customJar.jar")
+                customDependency.publicationJavadocPath = fileReference("customJar-javadoc.jar")
+                customDependency.publicationSourcePath = fileReference("customJar-sources.jar")
```

Comment:
This is ugly, but it's at least some way to specify javadoc and sources for the replacements. I haven't seen an easy way to get javadoc and sources artifacts from gradle - especially considering that these artifacts may be from non-default configurations.
I'm missing a good way to get from the `ProjectDependency` to the gradle `Project` here. Is there one / would it make sense to introduce that as a property on the `ProjectDependency`?

Reply:
> I'm missing a good way to get from the ProjectDependency to the gradle Project here. 

This is deliberate, you shouldn't be able to reach into projects through the resolution engine, projects should be isolated from each other. If we really wanted the javadoc/sources, we should make them variants of the project that we can query for, just like RUNTIME_JARS.

I'd leave this out for now and maybe give it a shot in a different PR.